## Lab 03.01 - Introduction to Pandas and Data Operations
In this lab, you'll learn how to use the pandas library to analyze real-world sales data. By the end of this lab, you'll be comfortable with the fundamental operations in pandas.

### Part 0 - Intro to Pandas
Before we dive into using pandas, let's understand what it is and why it's important in data science.

To answer the questions, edit the markdown cell and put your answer below the question. 

**Make sure to save the markdown cell, by pressing the ✓ (check) icon in the top right after answering the questions**

Research and critically think about the following questions:

##### Question 00
What is Pandas, and what are its primary functions in data visualization for data science?
- **Answer:** Pandas is a python library used for working with data sets. It has functions for analyzing, cleaning, exploring and manipulating data.  

##### Question 01
What does "data manipulation" mean in the context of pandas and data science field?
- **Answer:** Data manipulation means to manipulate your data by handling missing value and aranging a set of data to make it more organized and easier to intepret. 

##### Question 02
What kind of files can pandas read? List at least 3:
- **Answer:** the type of files that pandas can read are sql, csv, and excel. 

#### 0.0 - Hands-On Exploration
Let's start by importing pandas and exploring its basic functionality. Type these commands and write down what you observe:

First, let's install the matplotlib library, in your terminal run the following command:

```bash
pip3 install pandas
```

Once installed, let's start by importing pandas and exploring its basic functionality. 

In [10]:
import pandas as pd

Now lets run create a list, and see what happens when we use it with pandas.

In [12]:
# Create a simple list of numbers
numbers = [1, 2, 3, 4, 5]

# Convert it to a pandas Series
series = pd.Series(numbers)

# Print the result
print(series)

0    1
1    2
2    3
3    4
4    5
dtype: int64


##### Question 03
What did you notice about the output format?
- **Answer:** I noticed that there are 2 rows of numbers with one starting at 0 to 4 and another one from 1 to 5.

##### Question 04
How is this different from a regular Python list? What are the numbers on the left side?
- **Answer:** Ussually in a regular python list the numbers are next to each other going from left to right instead of up and down. The number on the left side are the series of the list.  

### Part 1 - Reading and Exploring Data

#### 01.00 - Loading Data
Let's start by loading our sales data that you downloaded alongside this lab. In pandas, we can read various file formats, but CSV (Comma-Separated Values) is one of the most common.

In [13]:
sales_df = pd.read_csv('sales_data.csv')

When using pandas to load data, in ingests it into what is known as a "dataframe", this allows us to do advanced manipulation.

#### 01.01 - Viewing Data Samples
Let's explore the fundamental methods for understanding our data structure and content. We'll look at each method individually and understand what it tells us about our data.

In [14]:
print(sales_df.head(5))

         Date    Product         Category    Price  Units    Total  Discount  \
0  2023-01-01   Notebook  Office Supplies    18.33      2    35.56      0.03   
1  2023-01-01     Laptop      Electronics  1943.35      4  7773.40      0.00   
2  2023-01-01     Tablet      Electronics   436.05      2   872.10      0.00   
3  2023-01-01  USB Drive      Accessories   116.48      3   349.44      0.00   
4  2023-01-01    Pen Set  Office Supplies    42.14      4   168.56      0.00   

  Region      Sales_Rep Payment_Method Customer_Segment  
0  South   Sarah Wilson     Debit Card   Small Business  
1  North     John Smith     Debit Card   Small Business  
2   East    David Brown  Bank Transfer       Government  
3   East  Lisa Anderson         PayPal       Individual  
4   East    David Brown     Debit Card        Corporate  


In [15]:
print(sales_df.tail())

            Date       Product         Category    Price  Units    Total  \
3766  2023-12-31    Smartwatch      Electronics  1730.72      2  3461.44   
3767  2023-12-31          Desk        Furniture   462.57      1   462.57   
3768  2023-12-31        Laptop      Electronics   560.79      2  1121.58   
3769  2023-12-31  Office Chair        Furniture   116.04      2   232.08   
3770  2023-12-31      Notebook  Office Supplies    18.15      4    72.60   

      Discount Region        Sales_Rep Payment_Method Customer_Segment  
3766       0.0  South  Michael Johnson     Debit Card       Individual  
3767       0.0  North       John Smith     Debit Card       Government  
3768       0.0   East      David Brown     Debit Card   Small Business  
3769       0.0  North       Emma Davis    Credit Card        Corporate  
3770       0.0  North       Emma Davis         PayPal       Individual  


##### Question 04
What is the difference between the `tail()` and `head()` commands? What would be use of either command be?
- **Answer:** The difference between the tail and head command is that the head function displays the first three rows of a data frame and the tail function is used to extract a specified number of items from the end of a sequence. Both the functions helps get a quick overview of a data content. 

##### Question 05
What happens when you put a number in the `head()` function? What changed?
- **Answer:** When you input a number it represents the index so if you were to put 5 the row will start at 0 and end at 4 showing 5 rows of data. 

#### 01.02 - Understanding DataFrame Structure
Let's examine the basic properties of our DataFrame:

In [16]:
print(sales_df.columns)

Index(['Date', 'Product', 'Category', 'Price', 'Units', 'Total', 'Discount',
       'Region', 'Sales_Rep', 'Payment_Method', 'Customer_Segment'],
      dtype='object')


In [17]:
print(sales_df.dtypes)

Date                 object
Product              object
Category             object
Price               float64
Units                 int64
Total               float64
Discount            float64
Region               object
Sales_Rep            object
Payment_Method       object
Customer_Segment     object
dtype: object


##### Question 06
Compare and contrast the outputs of the `columns` and `dtypes` properties. What is similar what is different?
- **Answer:** For the outputs of column it shows the different catagories for each column such as the date and product. The dtypes property shows each of the column catagory and the types of input in it such as wheter it will be a float etc. 

##### Question 07
What data type is the 'Date' column? Is this what you expected? 
- **Answer:** The data type of date column is object. This surprises me since I thought the data type for the date would have been datetime. 

##### Question 08
Why might pandas choose different data types for different columns?
- **Answer:** Pandas picks different data types for each column based on the kind of data in it to save memory and make calculations faster.

#### 01.03 - Data Information Summary
The `info()` method provides a concise summary of our DataFrame:

In [18]:
print(sales_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3771 entries, 0 to 3770
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date              3771 non-null   object 
 1   Product           3771 non-null   object 
 2   Category          3771 non-null   object 
 3   Price             3771 non-null   float64
 4   Units             3771 non-null   int64  
 5   Total             3771 non-null   float64
 6   Discount          3771 non-null   float64
 7   Region            3771 non-null   object 
 8   Sales_Rep         3693 non-null   object 
 9   Payment_Method    3739 non-null   object 
 10  Customer_Segment  3771 non-null   object 
dtypes: float64(3), int64(1), object(7)
memory usage: 324.2+ KB
None


##### Question 09
What information does `info()` tell us, how could this information be useful?
- **Answer:** The info function tells us the total number of colums, the data types, the memory ranges, the non-null count, and the range index. This information can be useful since it could help you understand your data as to wheter your missing something important. 

##### Question 10
How much memory is our DataFrame using? 
- **Answer:** Our data frame is using 324.2+ KB. 

##### Question 11
What does "null" mean? What impact could having "null" values in a dataset?
- **Answer:** Null values represent missing or undefined data in a dataset, and they can cause issues like biased results, reduced analysis accuracy, and performance slowdowns if not handled properly.

#### 01.04 - Numerical Summaries
The `describe()` method provides statistical summaries for numerical columns:

In [19]:
print(sales_df.describe())

             Price        Units         Total     Discount
count  3771.000000  3771.000000   3771.000000  3771.000000
mean    397.093911     2.985150   1159.701854     0.021594
std     514.524434     1.410003   1750.424109     0.041496
min       4.810000     1.000000      4.990000     0.000000
25%      40.165000     2.000000    110.730000     0.000000
50%     134.860000     3.000000    372.870000     0.000000
75%     516.490000     4.000000   1472.525000     0.020000
max    2073.680000     5.000000  10151.100000     0.150000


##### Question 12
What statistics does the `describe()` function give us?
- **Answer:** the statistics the describe function gives us is the count, mean, std, minimum, 25%, 50%, 75%, and maximum. 

##### Question 13
Which columns did the `describe()` function perform statistics on? Why? (Hint: why didnt we see 'Product' in the output?)
- **Answer:** the describe function only perform statistics on numeric and integer floating points, so therefore non numeric columns such as strings are excluded from this.   

In [51]:
print(sales_df[['Price', 'Units']].describe())

             Price        Units
count  3771.000000  3771.000000
mean    397.093911     2.985150
std     514.524434     1.410003
min       4.810000     1.000000
25%      40.165000     2.000000
50%     134.860000     3.000000
75%     516.490000     4.000000
max    2073.680000     5.000000


##### Question 14
What is different in the codeblock above compared to the previous one. How did this affect the output?
- **Answer:** The difference between the codeblock above and this one that here there is something inside the describe function with price and units so this function is only being used in these ones, compared to the one above who had way more columns. 

In [22]:
print(sales_df['Category'].unique())

['Office Supplies' 'Electronics' 'Accessories' 'Furniture']


##### Question 15
How many unique product categories are in our data set?
- **Answer:** There are 4 unique catagories in our data set. 

In [23]:
print("Output 1:")
print(sales_df['Category'].value_counts())

print('')
print('')

print("Output 2:")
print(sales_df['Category'].value_counts(normalize=True))

Output 1:
Category
Office Supplies    953
Accessories        942
Furniture          941
Electronics        935
Name: count, dtype: int64


Output 2:
Category
Office Supplies    0.252718
Accessories        0.249801
Furniture          0.249536
Electronics        0.247945
Name: proportion, dtype: float64


##### Question 16
What is the difference between the two outputs?
- **Answer:** The difference between the two outputs is that the first output data type is in integer while the second output data type is in a float. 

##### Question 17
What is our most common product category?
- **Answer:** Our most common product catagory is Office supplies. 

##### Excercise 00
Using what you've learned, create a code cell direclty below this one to answer these questions:

- What is the date range of our sales data? (Hint: try `min()` and `max()` functions)

     The date range of our sales is from january 1st 2023 to december 21st 2023.
      
- How many unique products do we sell?

      There are 20 unique products that we sell.

- What is our most common payment method?

      The most common payment method is a debit card. 

In [50]:
sales_df['Date'] = pd.to_datetime(sales_df['Date'])

# Now find the min and max dates
earliest_date = sales_df['Date'].min()
latest_date = sales_df['Date'].max()

print(f"Sales data ranges from {earliest_date} to {latest_date}")

Sales data ranges from 2023-01-01 00:00:00 to 2023-12-31 00:00:00


### Part 2 - Data Selection and Filtering

#### 02.01 - Column Selection
There are multiple ways to select columns in pandas. Let's explore them:

In [24]:
# Select single column
prices = sales_df['Price']

# Select multiple columns
product_info = sales_df[['Product', 'Price', 'Units']]

##### Question 17
Compare and contrast how we select multiple columns vs a single column.
- **Answer:** They both use sales_df in order to select the column that they want, and they both as well put what columns they want to show. The difference between the one with multiple columns is it has more than one string in the column compared to a single column which only has one. 

##### Excercise 01
Using what you've learned, create a code cell direclty below this one to select the 'Region' and 'Category' columns and store them in a variable called 'location_data'.

In [37]:
location_data = sales_df[['Region', 'Category']]
print(location_data)



     Region         Category
0     South  Office Supplies
1     North      Electronics
2      East      Electronics
3      East      Accessories
4      East  Office Supplies
...     ...              ...
3766  South      Electronics
3767  North        Furniture
3768   East      Electronics
3769  North        Furniture
3770  North  Office Supplies

[3771 rows x 2 columns]


#### 02.01 - Filtering Data
Let's learn how to filter our data based on conditions:

In [38]:
expensive_items = sales_df[sales_df['Price'] > 500]

north_sales = sales_df[sales_df['Region'] == 'North']

print("Expensive items:")
print(expensive_items)
print("\nNorth region sales:")
print(north_sales)

Expensive items:
            Date     Product     Category    Price  Units    Total  Discount  \
1     2023-01-01      Laptop  Electronics  1943.35      4  7773.40      0.00   
9     2023-01-01  Smartwatch  Electronics   893.71      3  2681.13      0.00   
10    2023-01-01      Laptop  Electronics   802.12      3  2093.53      0.13   
13    2023-01-02  Smartphone  Electronics   687.98      3  2063.94      0.00   
16    2023-01-02  Smartwatch  Electronics   507.91      4  2031.64      0.00   
...          ...         ...          ...      ...    ...      ...       ...   
3760  2023-12-31  Smartphone  Electronics  1766.20      4  7064.80      0.00   
3763  2023-12-31      Tablet  Electronics  1424.20      5  7121.00      0.00   
3764  2023-12-31  Smartphone  Electronics   857.34      5  4286.70      0.00   
3766  2023-12-31  Smartwatch  Electronics  1730.72      2  3461.44      0.00   
3768  2023-12-31      Laptop  Electronics   560.79      2  1121.58      0.00   

     Region          S

##### Question 18
What is `sales_df['Price'] > 500` accomplishing in our code? How might we use this in data science workflows?
- **Answer:** This function is displaying the datas that are expensive since their price is more than 500. This helps data scientist filter the data and gain valuable insights based on what they see. 

##### Question 19
What logic operator would we use to combine multiple conditonals when filtering columns?
- **Answer:** The logic operator we would use to combine multiple conditionals when filtering columns is and, or, not. 

##### Excercise 02
Using what you've learned, create a code cell direclty below this one to create the following filters:
- The number of units sold was greater than 5
- The product category is 'Electronics'

In [43]:
number_items = sales_df[sales_df['Units'] > 5]

electronic_sales = sales_df[sales_df['Category'] == 'Electronics'] 

print("expensive Units:")
print(number_items)
print("\nelectronics:")
print(electronic_sales)

expensive Units:
Empty DataFrame
Columns: [Date, Product, Category, Price, Units, Total, Discount, Region, Sales_Rep, Payment_Method, Customer_Segment]
Index: []

electronics:
            Date     Product     Category    Price  Units    Total  Discount  \
1     2023-01-01      Laptop  Electronics  1943.35      4  7773.40      0.00   
2     2023-01-01      Tablet  Electronics   436.05      2   872.10      0.00   
9     2023-01-01  Smartwatch  Electronics   893.71      3  2681.13      0.00   
10    2023-01-01      Laptop  Electronics   802.12      3  2093.53      0.13   
13    2023-01-02  Smartphone  Electronics   687.98      3  2063.94      0.00   
...          ...         ...          ...      ...    ...      ...       ...   
3760  2023-12-31  Smartphone  Electronics  1766.20      4  7064.80      0.00   
3763  2023-12-31      Tablet  Electronics  1424.20      5  7121.00      0.00   
3764  2023-12-31  Smartphone  Electronics   857.34      5  4286.70      0.00   
3766  2023-12-31  Smartw

#### Part 3 - Final Summary
Use your experience completing this lab, and the code block below to answer the following summary questions

##### Question 20
In your own words, explain what pandas is and why it's useful for data analysis:
- **Answer:** Panda is a part of python library where it is used for data manipulation and analysis. Its useful since it makes it easier for data analysis to find the specific data that they need and being able to filter it. 

##### Question 21
How many unique products are in the dataset?
- **Answer:**  there are 20 unique products in the data set. 

##### Question 22
What was the most popular product category in the `North` region? 
- **Answer:**  the most popular product catagory in the North region is furniture. 

My two codes based on groupby()  

In [67]:
#My first code 

sum = sales_df.groupby(["Product"])["Total"].sum()

# Print the result
print(sum)

Product
Bookshelf         190748.48
Desk              198208.10
Desk Organizer     13477.04
Filing Cabinet    203343.10
Headphones        668368.02
Keyboard           39767.49
Lamp              191325.75
Laptop            570923.24
Laptop Bag         51703.01
Monitor Stand      40499.19
Mouse              44730.82
Notebook           15726.38
Office Chair      151649.45
Paper Clips        16603.29
Pen Set            14513.18
Smartphone        601100.14
Smartwatch        638876.93
Stapler            15418.02
Tablet            658896.18
USB Drive          47357.88
Name: Total, dtype: float64


In [68]:
# My second code 

mean = sales_df.groupby(["Product","Units"])["Total"].mean()

print(mean) 


Product    Units
Bookshelf  1         352.800769
           2         683.100000
           3        1053.972353
           4        1372.044324
           5        2014.546970
                       ...     
USB Drive  1          82.493243
           2         188.317692
           3         227.497442
           4         290.947368
           5         413.406410
Name: Total, Length: 100, dtype: float64
